src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,218 @@ exprt java_bytecode_promotion(const exprt &expr)
+
+      // type is an atomic type (just one character)
+      else {
+        break;
+      }
+    }
+
+    std::string sub_str = src.substr(start, i - start + 1);
+    const typet &new_type = java_type_from_string(sub_str, class_name_prefix);
+    INVARIANT(new_type != nil_typet(), "Failed to parse type");
+
+    if (new_type.id() == ID_symbol) {
+      UNREACHABLE;
+    }
+
+    type_list.push_back(new_type);
+  }
+  return type_list;
+}
+
+/// For parsing a class type reference
+/// \param src: The input string
+///   Either a signature: "Lpackage/class<TT;>.innerclass<Ljava/lang/Foo;>;
+///   Or a descriptor: "Lpackage.class$inner;
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+/// \return The reference type if parsed correctly, no value if parsing fails.
+optionalt<reference_typet>
+build_class_name(const std::string src, const std::string &class_name_prefix) {
+  PRECONDITION(src[0] == 'L');
+  // ends on ;
+  if (src[src.size() - 1] != ';')
+    return optionalt<reference_typet>();
+
+  std::string container_class = gather_full_class_name(src);
+  std::string identifier = "java::" + container_class;
+  symbol_typet symbol_type(identifier);
+  symbol_type.set(ID_C_base_name, container_class);
+
+  std::size_t f_pos = src.find('<', 1);
+  if (f_pos != std::string::npos) {
+    java_generic_typet result(symbol_type);
+    // get generic type information
+    do {
+      std::size_t e_pos = find_closing_delimiter(src, f_pos, '<', '>');
+      if (e_pos == std::string::npos)
+        throw unsupported_java_class_signature_exceptiont("recursive generic");
+
+      add_generic_type_information(result, src.substr(f_pos, e_pos - f_pos + 1),
+                                   class_name_prefix);
+
+      // Look for the next generic type info (if it exists)
+      f_pos = src.find('<', e_pos + 1);
+
+    } while (f_pos != std::string::npos);
+    return result;
+  }
+
+  return java_reference_type(symbol_type);
+}
+
+/// Finds the closing semi-colon ending a ClassTypeSignature that starts at
+/// \p starting_point.
+/// \param src: The input string to work on.
+/// \param starting_point: The string position where the opening 'L' we want to
+///   find the closing ';' for.
+/// \return The string position corresponding to the matching ';'. For example:
+/// LA;, we would return 2. For LA<TT;>; we would return 7.
+/// See unit/java_bytecode/java_util_tests.cpp for more examples.
+size_t find_closing_semi_colon_for_reference_type(const std::string src,
+                                                  size_t starting_point) {
+  PRECONDITION(src[starting_point] == 'L');
+  size_t next_semi_colon = src.find(';', starting_point);
+  INVARIANT(next_semi_colon != std::string::npos,
+            "There must be a semi-colon somewhere in the input");
+  size_t next_angle_bracket = src.find('<', starting_point);
+
+  while (next_angle_bracket < next_semi_colon) {
```

Comment:
This might work, but I'd like to see some argument about why this exhaustively covers the type signature grammar... if this gets too tricky we should probably abandon manual parser writing and instead make a Bison or similar parse for the signature grammar (https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.4)

This exists but is GPL'd: https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/sun/reflect/generics/parser/SignatureParser.java

Reply:
Yes I completely agree, unfortunately there isn't time at this juncture to do this. Here is why I'm convinced this is reasonable for now:

This is for parsing a reference type from either a signature or descriptor. If it is descriptor, then the grammar is exactly:
`L ClassName ;`

So since ClassName can't contain `;` or `<` we're done. 

(Which is why it worked until now). 

For signatures, it is more complicated:
```
L PackageSpecifier_opt SimpleClassTypeSignature ClassTypeSignatureSuffix* ;
```

The package specifier is just `identifier/identifier` so no chances of `;` or `<'`

Then the `SimpleClassTypeSignature` is 

```
Identifier TypeArguments_opt
```
Where type arguments must start with `<` (so this is the case where `next_angle_bracket < next_semi_colon` and must end with a corresponding `>` 

This is then optionally followed by a `.` indicating inner class and then repeat above argument. It is finally ended by a semi colon. 

I appreciate this is _not_ convincing, hopefully the extensive unit tests are *more* convincing - though I accept they are a long way from an exhaustive proof.